enclave/qeidinfo.c

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/bits/safecrt.h>
+#include <openenclave/bits/safemath.h>
+#include <openenclave/enclave.h>
+#include <openenclave/internal/calls.h>
+#include <openenclave/internal/enclavelibc.h>
+#include <openenclave/internal/raise.h>
+#include "../common/qeidentity.h"
+
+#include <stdio.h>
+/**
+ * Validate and copy buffer to enclave memory.
+ */
+static oe_result_t _copy_buffer_to_enclave(
+    uint8_t** dst,
+    size_t* dst_size,
+    const uint8_t* src,
+    size_t src_size)
+{
+    oe_result_t result = OE_FAILURE;
+    if (!src || src_size == 0 || !oe_is_outside_enclave(src, src_size) ||
+        dst == NULL || dst_size == NULL)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    *dst = oe_malloc(src_size);
+    if (*dst == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    OE_CHECK(oe_memcpy_s(*dst, src_size, src, src_size));
+    *dst_size = src_size;
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+/**
+ * Call into host to fetch qe identity information.
+ */
+oe_result_t oe_get_qe_identity_info(oe_get_qe_identity_info_args_t* args)
+{
+    oe_result_t result = OE_FAILURE;
+    size_t host_args_buffer_size = sizeof(*args);
+    uint8_t* host_args_buffer = NULL;
+    oe_get_qe_identity_info_args_t* host_args = NULL;
+    oe_get_qe_identity_info_args_t tmp_args = {0};
+
+    printf("Enclave: Calling %s\n", __PRETTY_FUNCTION__);
+
+    if (args == NULL)
+        OE_RAISE(OE_FAILURE);
+
+    // allocate host memory to hold enclave arguments before calling into host
+    // for getting qe identity info
+    host_args_buffer = oe_host_malloc(host_args_buffer_size);
+    if (host_args_buffer == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    // Copy args struct.
+    host_args = (oe_get_qe_identity_info_args_t*)host_args_buffer;
+    *host_args = *args;
+
+    result = oe_ocall(OE_OCALL_GET_QE_ID_INFO, (uint64_t)host_args, NULL);
+    if ((result == OE_OK) &&
```

Comment:
Line 66-71 was added to catch OE_QUOTE_PROVIDER_CALL_ERROR error specifically so that we could propagate it back to the caller to decide whether QE Identity is available for QE checking.
Not available is not a error condition before the quote provider library is available with QE Identity feature.

We want to raise error on any Any other error codes

Reply:
As discussed, we could simplify this.

Reply:
I mis-understood your original comment.
You are right. Your suggested fix makes the code much cleaner.

Thanks a lot.